frontend/catalyst/utils/runtime.py

```diff
@@ -159,77 +170,125 @@ def is_not_ctrl(op):
     """Check that device.operations is equivalent to the union of *args
 
     Args:
-        device (qml.Device): An instance of a quantum device.
-        *args (List[Str]): List of strings.
+        device_gates (Set[str]): device gates
+        spec_gates (Set[str]): spec gates
 
     Raises: CompileError
     """
-    operations = filter_out_adjoint_and_control(device.operations)
-    gates_in_device = set(operations)
-    set_of_sets = [set(arg) for arg in args]
-    union = set.union(*set_of_sets)
-    if gates_in_device == union:
+    if device_gates == spec_gates:
         return
 
-    msg = "Gates in qml.device.operations and specification file do not match"
+    msg = (
+        "Gates in qml.device.operations and specification file do not match.\n"
+        f"Gates that present only in the device: {device_gates - spec_gates}\n"
+        f"Gates that present only in spec: {spec_gates - device_gates}\n"
+    )
     raise CompileError(msg)
 
 
-def check_gates_are_compatible_with_device(device, config):
-    """Validate configuration dictionary against device.
+def validate_config_with_device(device: qml.QubitDevice, config: TOMLDocument) -> None:
+    """Validate configuration document against the device attributes.
+    Raise CompileError in case of mismatch.
+    * if configuration file does not exists
+    * if decomposable, matrix, and native gates have some overlap
+    * if decomposable, matrix, and native gates do not match gates in ``device.operations``
 
     Args:
         device (qml.Device): An instance of a quantum device.
-        config (Dict[Str, Any]): Configuration dictionary
 
     Raises: CompileError
     """
 
-    native = get_native_gates(config)
-    decomposable = get_decomposable_gates(config)
-    matrix = get_matrix_decomposable_gates(config)
-    check_no_overlap(native, decomposable, matrix)
-    if not hasattr(device, "operations"):  # pragma: nocover
-        # The new device API has no "operations" field
-        # so we cannot check that there's an overlap or not.
-        return
+    if not config["compilation"]["qjit_compatible"]:
+        raise CompileError(
+            f"Attempting to compile program for incompatible device '{device.name}': "
+            f"Config is not marked as qjit-compatible"
+        )
 
-    check_full_overlap(device, native, decomposable, matrix)
+    shots_present = device.shots is not None
+    native = get_pennylane_operations(config, shots_present)
+    observables = set(get_observables(config, shots_present))
+    decomposable = set(get_decomposable_gates(config, shots_present))
+    matrix = set(get_matrix_decomposable_gates(config, shots_present))
+
+    # Filter-out ControlledQubitUnitary because some devices are known to support it.
+    # TODO: Should we ignore this ambiguity instead?
+    if "ControlledQubitUnitary" in native and check_quantum_control_flag(config):
+        matrix = matrix - {"ControlledQubitUnitary"}
```

Comment:
I have updated the comment, it should be clear now.

Reply:
Ah got it, I'm not sure it should be filtered out though. Doesn't it brake the verification if a device lists it under `device.operations` but it doesn't show up anywhere in the TOML?
I might be safest to just leave it alone (in the context of the old schema).

Reply:
But the verification would be failed for schema 1 because, namely our [overalp](https://github.com/PennyLaneAI/catalyst/blob/2974551a0b6f9a45fc1260b40115606e01bc9783/frontend/catalyst/utils/runtime.py#L256) condition. It is a pure-config integrity check. Device does list this op in `.operations` no problems with this.

Reply:
Why would it fail the overlap condition? The TOML should only place the gate in either one of native, matrix, decomp.

Reply:
In toml spec 1 we deduce the controled ops (quantum_control flag is True which is what actually happens), so we add this gate to the set of native gates. But the spec already contains it in the matrix/decomp section, thus the conflict.

Reply:
Where in the code are we adding this gate? If anything we would be adding `C(QubitUnitary)`, not `ControlledQubitUnitary` right?

Reply:
Discussed this offline we'll remove it after the TOML update in the lightning repo.